File: apps/mobile/android/app/src/main/kotlin/com/voicememory/mobile/HardwareMonitorHandler.kt
package com.voicememory.mobile

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build
import android.os.PowerManager
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

object HardwareMonitorHandler {
    fun handle(context: Context, call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getThermalStatus" -> result.success(readThermalStatus(context))
            "getHardwareSnapshot" -> result.success(readSnapshot(context))
            else -> result.notImplemented()
        }
    }

    private fun readSnapshot(context: Context): Map<String, Any> {
        val battery = context.registerReceiver(
            null,
            IntentFilter(Intent.ACTION_BATTERY_CHANGED),
        )
        val level = battery?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
        val scale = battery?.getIntExtra(BatteryManager.EXTRA_SCALE, 100) ?: 100
        val percent = if (level < 0 || scale <= 0) -1 else (level * 100) / scale
        val status = battery?.getIntExtra(BatteryManager.EXTRA_STATUS, -1) ?: -1
        val charging = status == BatteryManager.BATTERY_STATUS_CHARGING ||
            status == BatteryManager.BATTERY_STATUS_FULL
        return mapOf(
            "batteryPercent" to percent,
            "isCharging" to charging,
            "thermalStatus" to readThermalStatus(context),
        )
    }

    private fun readThermalStatus(context: Context): String {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
            return "unknown"
        }
        val powerManager =
            context.getSystemService(Context.POWER_SERVICE) as PowerManager
        return when (powerManager.currentThermalStatus) {
            PowerManager.THERMAL_STATUS_NONE -> "nominal"
            PowerManager.THERMAL_STATUS_LIGHT -> "fair"
            PowerManager.THERMAL_STATUS_MODERATE -> "moderate"
            PowerManager.THERMAL_STATUS_SEVERE -> "severe"
            PowerManager.THERMAL_STATUS_CRITICAL -> "critical"
            PowerManager.THERMAL_STATUS_EMERGENCY -> "critical"
            PowerManager.THERMAL_STATUS_SHUTDOWN -> "critical"
            else -> "unknown"
        }
    }
}

File: apps/mobile/ios/Runner/HardwareMonitorHandler.swift
import Flutter
import UIKit

/// Battery and thermal snapshot for on-device transcription and summaries.
enum HardwareMonitorHandler {
  static func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getThermalStatus":
      result(thermalStatus())
    case "getHardwareSnapshot":
      result(snapshot())
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  static func snapshot() -> [String: Any] {
    let device = UIDevice.current
    device.isBatteryMonitoringEnabled = true
    let level = device.batteryLevel
    let percent = level < 0 ? -1 : Int((level * 100).rounded())
    let charging = device.batteryState == .charging || device.batteryState == .full
    return [
      "batteryPercent": percent,
      "isCharging": charging,
      "thermalStatus": thermalStatus(),
    ]
  }

  static func thermalStatus() -> String {
    switch ProcessInfo.processInfo.thermalState {
    case .nominal:
      return "nominal"
    case .fair:
      return "fair"
    case .serious:
      return "serious"
    case .critical:
      return "critical"
    @unknown default:
      return "unknown"
    }
  }
}
